### PR TITLE
Add draft account-registration spec

### PR DESCRIPTION
**Pull request by progval**
This draft specification adds **simple** `REGISTER` and `VERIFY` commands to standardize account creation, as a replacement for the withdrawn [Account Management Framework](https://github.com/ircv3/ircv3-specifications/pull/276)

It is based on @edk0's [draft](https://gist.github.com/edk0/bf3b50fc219fd1bed1aa15d98bfb6495), and is implemented by Oragono and Limnoria.

[rendered view](https://github.com/ProgVal/ircv3-specifications/blob/register/extensions/register.md)

### CODE REVIEW

**Comment by westor7** [comment]
Very nice idea and i hope for this, having some quick questions about this:

1. What if the irc services having VERIFY disabled, so a user no need to verify his email?
2. What if a user due the VERIFY sends an incorrect code or incorrect account?
3. What if the irc services will be down?

- Thanks!

**Comment by ProgVal** [comment]
>     1. What if the irc services having VERIFY disabled, so a user no need to verify his email?

Then the server responds with `REGISTER SUCCESS` instad of `REGISTER VERIFICATION_REQUIRED`.

>     2. What if a user due the VERIFY sends an incorrect code or incorrect account?
>     3. What if the irc services will be down?

Good catch. Added response codes `FAIL VERIFY INVALID_CODE` and `FAIL REGISTER/VERIFY UNAVAILABLE`.

**Comment by westor7** [comment]
https://github.com/ircv3/ircv3-specifications/pull/435/commits/339361e863dba410c5bea12ec942c4acdc8a9514#diff-2050f0354e63e54a43299643306ae9c6b94a0a5e24af547c1b50a95fadbfc81aR122

You mean if NOT right?

So: ``Sent by the server if the `REGISTER`/`VERIFY` commands are temporarily available.``
Should be: ``Sent by the server if the `REGISTER`/`VERIFY` commands are temporarily unavailable.``

**Comment by DanielOaks** [comment]
We should reach out to see whether Atheme+Anope have any thoughts on this. Any ircds have thoughts about it? (obviously Ora likes it, and we started using the `before-connect` key because that supports certain kinds of restricted installs that we support but I don't think traditional irc nets would be into that aspect).

**Comment by aaronmdjones** [comment]
Why do the success messages resemble e.g. `REGISTER SUCCESS` but the failure messages resemble `FAIL REGISTER`? Should they not be the other way around?

**Comment by ProgVal** [comment]
@aaronmdjones The error messages follow the new [Standard Replies Extension](https://ircv3.net/specs/extensions/standard-replies)

File: extensions/register.md
Comment (hhirtz) [inline extensions/register.md:46]: Thoughts on

```suggestion
The defined keys are:

 * `before-connect` - if present, indicates the server supports early registration, so it
   MUST NOT use `FAIL REGISTER COMPLETE_CONNECTION_REQUIRED`
 * `verify` - if present, registrations require verification to proceed
 * `verify-email` - if present, verification can be done via email
 * `verify-tel` - if present, verification can be done via telephone

### Commands

    REGISTER {verify-mean=<verify-value> | "*"} <password>
```

?

It'd allow something like

```
S: CAP * LS :register=before-connect,verify,verify-email,verify-tel
C: REGISTER tel=+33612345678 :a very secure password
```

Vendor verification means could be added like so:

```
S: CAP * LS :register=before-connect,verify,verify-email,vendor/verify-matrix
C: REGISTER vendor/matrix=@freenode_pinkieval:matrix.org :though this should be strictly checked
```

Comment (emersion) [reply extensions/register.md:46]: I don't think we should try to make this spec as generic as ACC was.

Comment (ProgVal) [reply extensions/register.md:46]: I agree with @emersion. I'll add a requirement that clients should ignore unknown keys, so future specification can extend this one by adding keys.

**Comment by edk0** [comment]
null characters can't be in IRC messages anyway so mentioning them feels a bit redundant

File: extensions/register.md
Comment (DanielOaks) [inline extensions/register.md:112]: You'll need the `<message>` param on the end of these to match the std replies format.

File: extensions/register.md
Comment (slingamn) [inline extensions/register.md:134]: Pure pedantry: I don't think SASL PLAIN would have any problems here, since everything gets base64-encoded. The issue is more with being able to pass them easily to NickServ (e.g., to change the password).

I think a warning about password length may be in order? I've been imposing a cap of 300 bytes to ensure that passwords fit comfortably into IRC lines (it's theoretically possible that someone could submit a passphrase through this framework that gets truncated in other contexts). And although I haven't been validating this until now, it might also be worth checking that passphrases are acceptable as non-final IRC parameters (nonempty, no spaces, no initial `:`).

Comment (ProgVal) [reply extensions/register.md:134]: Purer pedantry: RFC 4616 requires the encoded bytes to be a UTF-8 string.

IMO warnings about password length should be left to the UI. I could mention it in a non-normative section though.

**Comment by DanielOaks** [comment]
So can we put this on the website and wait for implementations?

**Comment by slingamn** [comment]
I'll compound my embarrassment a bit by requesting a "security considerations" section reflecting https://github.com/oragono/oragono/issues/1634 --- something like:

>Servers MUST ensure that if they allow `REGISTER` before connection registration, this functionality cannot be used to bypass any authentication restrictions defined in the server configuration, e.g. requirements that clients send `PASS` or authenticate with SASL.

**Comment by SadieCat** [comment]
Trivial suggestion (don't block merging on this): Might be a good idea to rename the spec to "account registration" rather than just "register".

File: extensions/register.md
Comment (emersion) [inline extensions/register.md:45]: In case the server supports `before-connect`, what is the "curerent nick"? Should we require that a `NICK` command is sent before `REGISTER`?

File: extensions/register.md
Comment (DanielOaks) [inline extensions/register.md:89]: s/is/if/

File: extensions/register.md
Comment (DanielOaks) [reply extensions/register.md:45]: nice catch. something like this in the list of responses should be sufficient? and/or a line on the `before-connect` key description. it's possible in `before-connect` setups but given those are mostly an experimental use case tossed on top I'd rather not describe this in the 'mainline' spec path.

>     FAIL REGISTER NEED_NICK <message>
> 
> Sent by the server if the client sends `REGISTER` before sending a `NICK` command. This is only possible in setups that advertise the `before-connect` key.

**Comment by SadieCat** [comment]
It might be a good idea to have password min/max lengths in the register cap value too?

File: extensions/register.md
Comment (SadieCat) [inline extensions/register.md:74]:
```suggestion
    FAIL REGISTER ACCOUNT_EXISTS <account> <message>
```

Would prevent overloaded terminology (username can also refer to what is also known as the ident).

File: extensions/register.md
Comment (SadieCat) [inline extensions/register.md:39]: Should there be a fail code for when the user sends * instead of an email or is INVALID_EMAIL enough?

Comment (DanielOaks) [reply extensions/register.md:39]: INVALID_EMAIL should be enough, the client software can work it out based on that and the message should tell the user what went wrong

**Comment by emersion** [comment]
A WIP client implementation is available for gamja in this branch: https://git.sr.ht/~emersion/gamja/log/register-account

Should be enough to test basic functionality.

**Comment by DanielOaks** [comment]
Do we need Atheme/Anope to look at this or is this something that ircds will implement all on their own without services involvement? If we do need services to look at this who from those projects can look over this and give us the yay/nay on it?

**Comment by SadieCat** [comment]
Services will need per-ircd support like with SASL probably (I'm implementing the Anope side of it).

File: extensions/register.md
Comment (ProgVal) [reply extensions/register.md]: I added a `*` argument before `<message>`, for consistency with the other errors.

**Comment by ProgVal** [comment]
I applied all the suggestions above. In particular, note that the capability was renamed `draft/account-registration` instead of `draft/register`, for consistency with this spec's name, and other capability names.

File: extensions/account-registration.md
Comment (kaniini) [inline extensions/account-registration.md:59]: For Ophion, and other environments where accounts exist, but nicknames are not owned (like Undernet for example), this is somewhat awkward.

It would be nice if we could have it be:

    REGISTER {<email> | "*"} [accountname] <password>

This would work for Ophion (and presumably cases like Undernet too).

File: extensions/account-registration.md
Comment (kaniini) [inline extensions/account-registration.md:70]: Not all environments want to send email, for example, Ophion intends to prompt the user to complete a CAPTCHA it generates at some URI.  Changing "email verification" to "verification" would allow more flexibility.

**Review by kaniini — CHANGES_REQUESTED** [review]

File: extensions/account-registration.md
Comment (kaniini) [inline extensions/account-registration.md:106]: The `ALREADY_REGISTERED` fail code is sent if a client is already *authenticated*.  Should this be `ALREADY_AUTHENTICATED`?

Comment (DanielOaks) [reply extensions/account-registration.md:106]: makes sense

File: extensions/account-registration.md
Comment (ProgVal) [reply extensions/account-registration.md:61]: Done by @DanielOaks through a PR to my repo.

File: extensions/account-registration.md
Comment (DanielOaks) [inline extensions/account-registration.md:59]: this should probably be left as accountname so that it's more clear

Comment (ProgVal) [reply extensions/account-registration.md:59]: The rest of the spec uses `<account>`.

File: extensions/account-registration.md
Comment (SadieCat) [inline extensions/account-registration.md:194]:
```suggestion
like a `PRIVMSG` to NickServ), which are limited in length, servers may want to
```

File: extensions/account-registration.md
Comment (SadieCat) [inline extensions/account-registration.md:181]:
```suggestion
like a `PRIVMSG` to NickServ), which are limited in length, clients may want to
```

File: extensions/account-registration.md
Comment (DanielOaks) [inline extensions/account-registration.md:156]: oop, I forgot to update this to the new cap name `draft/account-registration`

**Comment by slingamn** [comment]
Can we keep the draft name as `draft/register`, then plan to verify it as `account-registration`? (Same strategy as chathistory: we're keeping the draft name as `draft/event-playback` and planning to ratify as `chathistory-events`.)

**Comment by progval** [comment]
Can we merge this?

**Comment by jwheare** [comment]
@SadieCat is this implemented in Anope now?

**Comment by SadieCat** [comment]
I do have a WIP implementation. I'll try to get it onto the InspIRCd Testnet this week.

**Comment by progval** [comment]
@SadieCat Any news?

**Comment by jwheare** [comment]
Probably good to merge this as draft even without an anope implementation. Client devs can test using ergo. Can do so next week unless there's any blockers.

**Comment by jwheare** [comment]
Blocking merge pending resolution on ircv3/ircv3-ideas#78

**Comment by jwheare** [comment]
Dropping draft/ isn't clear cut so unblocking this.

**Comment by jwheare** [comment]
The motivation section probably doesn't need to mention the history of ACC or go into how the authors hope it will be acceptable. It can just be more direct about what the spec is for on its own terms. @progval can you have a stab at fixing that? Otherwise I can merge and we'll fix it before ratifying.

File: extensions/account-registration.md
Comment (jwheare) [inline extensions/account-registration.md:55]: This needs to be moved into the usual "Notes for implementing work-in-progress version" style at the start. See e.g. #466 instead of burying it in another section.

File: extensions/account-registration.md
Comment (jwheare) [inline extensions/account-registration.md:36]: Not sure we need to mention this precedent here. The format should already be described in a standalone way.

File: extensions/account-registration.md
Comment (jwheare) [inline extensions/account-registration.md:178]: Fix the line breaking here (and elsewhere). Either hard wrap for line length or (more preferably) soft wrap, hard wrapping on punctuation is odd.

File: extensions/account-registration.md
Comment (jwheare) [inline extensions/account-registration.md:205]: Can remove this when reference is dropped further up.

**Review by jwheare — CHANGES_REQUESTED** [review]
Actually a few quick things I'd like changing before merge.

File: extensions/account-registration.md
Comment (progval) [reply extensions/account-registration.md:36]: I don't understand; do you want me to copy-paste it here?

Comment (emersion) [reply extensions/account-registration.md:47]: I think @jwheare wants to drop this part "the format is intended to follow the precedent set by multiline", while keeping the rest of the sentence.

Comment (progval) [reply extensions/account-registration.md:47]: "The capability has an optional value, a comma-separated list of key-value pairs" alone is not enough to describe the format.

Comment (jwheare) [reply extensions/account-registration.md:47]: If you need to copy paste it here to make it standalone then yeah.

**Comment by progval** [comment]
All good now?

File: extensions/account-registration.md
Comment (jwheare) [inline extensions/account-registration.md:35]: How about this? Historically, other specs lean on the word "historically" in this section.

```suggestion
A standard method of registering a user account has historically been missing
from IRC. Instead various methods such as sending messages to services bots
(e.g. NickServ) have been employed.

This specification aims to provide a standard baseline for account
registration that allows client developers to build more reliable and
integrated user experiences around account registration.

It does not aim to cover all registration-related use cases that may currently
be in use by server implementations to varying degrees. However, future
extensions to the specification might be proposed given a broad enough
consensus.
```

Comment (progval) [reply extensions/account-registration.md]: Nice!

**Review by jwheare — APPROVED** [review]
Alright looks good to merge now.

**Pull request merged**
